pyvisa-py/usb.py

```diff
@@ -194,14 +197,40 @@ def list_resources():
                                   usb_interface_number=intfc))
         return out
 
+    def read(self, count):
```

Comment:
Doing this bypass the whole logic in https://github.com/pyvisa/pyvisa-py/blob/master/pyvisa-py/sessions.py#L436 which I do not really like. Furthermore you simply swallow all usb exceptions that do not match a timeout. I think the proper way to do would be to provide to `self._read` a custom reader function handling the exception and a custom exception class used to differentiate between a timeout and something else. You could define a method/property returning the proper reader function. By default it would simply return `lambda: self.interface.read(count)`, and for USBInstr it would be:
``` python 
def _usb_reader(self):
    """Data reader identifying usb timeout exception.

    """
    try:
        data = self.interface.read(count)
    except usb.USBError as exc:
        if exc.errno in (errno.ETIMEDOUT, -errno.ETIMEDOUT):
            raise USBTimeoutException()
        raise
```
Furthermore it looks to me like this can apply also to all usb session not only USBTMC, which means you can just update read and use a single reader function.

Reply:
Swallowing non-timeout exceptions was an unintentional mistake in my code. I'm glad you caught it.

I opted to bypass Session._read() because it calls the low-level read function repeatedly in a loop, until the result is non-empty. This may be appropriate for raw USB instruments, where you just read until some response or timeout. But USBTMC communication is explicitly message-based, and a device could in principle send an empty response. Repeating the read primitive would not be appropriate in this case.

This is admittedly an edge-case. I think your version would work well enough in practice.

Reply:
I do agree that for a system implementing USBTMC only it would not make sense, but here we are considering system building the VISA layer on top of USBTMC. And generally in VISA when ones performs a read operation the instrument is expected to answer something. Even if the answer is empty I would expect to see the termination character.
Because of this, I still believe it worthwhile to go through the `_read` method to enforce the VISA specification concerning the rule 6.1.

Reply:
Ok, understood. I implemented your approach with the _usb_reader() function.